docs/features/DefaultInterfaceImplementation.md

```diff
@@ -43,16 +43,41 @@ class Test1 : I1
 
 - Declaring types within interfaces. **Protected** and **protected internal** accessibility is not supported.
 
-- Implementing interface methods in derived interfaces by using explicit implementation syntax, accessibility is private, allowed modifiers: **extern** and **async**.
+- Implementing interface methods in derived interfaces by using explicit implementation syntax, accessibility matches accessibility of the implemented member, allowed modifiers: **extern** and **async**.
```

Comment:
>matches accessibility of the implemented member [](start = 110, length = 47)

This doesn't ensure that the member is accessible to the caller.  For example, an `internal` implementation in one assembly can be implemented in another assembly with internals-visible-to access to the first.  A third assembly with internal-visible-to the original interface but not the override's assembly would not have access to that override.  To ensure the override is accessible to the caller (and nowhere else), explicit override methods should be given `protected` access.

Reply:
> This doesn't ensure that the member is accessible to the caller. 

I am not claiming that it is.

> To ensure the override is accessible to the caller (and nowhere else), explicit override methods should be given protected access.

1. In general, giving protected access would be incorrect because that might expose otherwise inaccessible types.
2. So far, we disallow declaring protected members in interfaces and haven't defined what protected access means for interface members both for the language and for the runtime.

That is why I am starting with the same accessibility and we will refine this once we discuss design issues at LDM


---
In reply to: [255757955](https://github.com/dotnet/roslyn/pull/33281#discussion_r255757955) [](ancestors = 255757955)

Reply:
Can you please explain how declaring a protected member in IL (that has an unutterable name) could "expose" an inaccesible type?  I'm assuming we would not change the accessibility of types just because we emit symbols that use them.

---
In reply to: [255762670](https://github.com/dotnet/roslyn/pull/33281#discussion_r255762670) [](ancestors = 255762670,255757955)

Reply:
> Can you please explain how declaring a protected member in IL (that has an unutterable name) could "expose" an inaccesible type?  I'm assuming we would not change the accessibility of types just because we emit symbols that use them.

The name doesn't matter in my opinion. Original member declaration can be internal and, therefore, refer to internal types. It would be illegal to use those types in a protected member.  

---
In reply to: [255770151](https://github.com/dotnet/roslyn/pull/33281#discussion_r255770151) [](ancestors = 255770151,255762670,255757955)

Reply:
It is possible that protected access would work, if we new what that means. As I said, this is a topic for an LDM.

---
In reply to: [255770872](https://github.com/dotnet/roslyn/pull/33281#discussion_r255770872) [](ancestors = 255770872,255770151,255762670,255757955)

Reply:
> Original member declaration can be internal and, therefore, refer to internal types. It would be illegal to use those types in a protected member.  

I'm not aware of an IL restriction that prevents using internal types in a protected member.  I don't understand any language design issue to be impacted by this; what question would the LDM be deciding?  The only issue is whether the proposed implementation in this PR is correct, and you're saying "I am not claiming that it is".  Since we want a correct implementation, I think it needs to be revised.

---
In reply to: [255772556](https://github.com/dotnet/roslyn/pull/33281#discussion_r255772556) [](ancestors = 255772556,255770872,255770151,255762670,255757955)

Reply:
Ah, I understand: what is the meaning of a `protected` member in an interface.  Added to open issues.

---
In reply to: [255803140](https://github.com/dotnet/roslyn/pull/33281#discussion_r255803140) [](ancestors = 255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> I'm not aware of an IL restriction that prevents using internal types in a protected member.  I don't understand any language design issue to be impacted by this; what question would the LDM be deciding? 

Try to compile this code, for example:
```
internal class X { }

public class Y
{
    protected X P { get; }
}
```

Compiler reports an error:
```
Error CS0053 Inconsistent accessibility: property type 'X' is less accessible than property 'Y.P' 
```

Also, note that the approach to match accessibility rather than change it is not worse than with a regular overriding. With regular overriding in a class, accessibility should match and that can create problems with internal virtual members too. See [https://github.com/dotnet/roslyn/issues/33329](https://github.com/dotnet/roslyn/issues/33329).


> The only issue is whether the proposed implementation in this PR is correct, and you're saying "I am not claiming that it is".  

I am saying that I am not claiming that the implementation in derived interface is always going to be accessible, however it has good chances to be accessible in majority of scenarios. I also believe that this approach is reasonable and consistent with regular overriding in classes. I am not sure that changing accessibility to protected would be correct (given the simple example above). So, I've made a call and, for the first stub at base access feature, I defined what is correct. I am happy to discuss this at LDM and change the implementation accordingly once we make a decision.


> Since we want a correct implementation, I think it needs to be revised.

I am claiming that the current implementation is correct according to the speclet and not planning to make an adjustment until LDM makes a call.  

---
In reply to: [255803908](https://github.com/dotnet/roslyn/pull/33281#discussion_r255803908) [](ancestors = 255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> Try to compile this code

You're demonstrating a C# source language restriction.  The generated code does not have to be expressible in source.  I'm not aware of any corresponding IL restriction.

In any case I've placed it on the open issue list and queued it for LDM discussion.

---
In reply to: [256177815](https://github.com/dotnet/roslyn/pull/33281#discussion_r256177815) [](ancestors = 256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> I am claiming that the current implementation is correct according to the speclet

Yes, this PR updates the speclet and the implementation to match each other, though it produces cases that require that we generate illegal IL.

---
In reply to: [256257340](https://github.com/dotnet/roslyn/pull/33281#discussion_r256257340) [](ancestors = 256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> Yes, this PR updates the speclet and the implementation to match each other, though it produces cases that require that we generate illegal IL.

Could you elaborate, give an example please? 

---
In reply to: [256257520](https://github.com/dotnet/roslyn/pull/33281#discussion_r256257520) [](ancestors = 256257520,256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> > Yes, this PR updates the speclet and the implementation to match each other, though it produces cases that require that we generate illegal IL.

> Could you elaborate, give an example please?

Yes, the example is
``` c#
// assembly1
[InternalsVisibleTo("assembly2")]
[InternalsVisibleTo("assembly3")]
public interface I1
{
    internal void M() {}
}
```
``` c#
// assembly2
public interface I2 : i1
{
    void I1.M() {}
}
```
``` c#
// assembly3
public interface I3: I1, I2
{
    void I1.M()
    {
        // this call must emit an invocation of I2.M
        // but that is not CLR-accessible in the speclet's translation scheme
        base(I2).M();
    }
}
```


---
In reply to: [256265573](https://github.com/dotnet/roslyn/pull/33281#discussion_r256265573) [](ancestors = 256265573,256257520,256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> An example is
> ...
>
>      // this call must emit an invocation of I2.M
>      // but that is not CLR-accessible in the speclet's translation scheme
>       base(I2).M();

I think this is misinterpretation of the speclet and misinterpretation of the implementation:
- "The most specific implementation must be accessible at the call site." This is a requirement for the base access that should be met, not a guarantee.
- According to the above, compiler shouldn't emit a call to the inaccessible method in the example, but should report an error. Unit-test ExplicitBase_152 clearly demonstrates this behavior.

Do you still believe that the speclet "produces cases that require that we generate illegal IL"?


---
In reply to: [256314854](https://github.com/dotnet/roslyn/pull/33281#discussion_r256314854) [](ancestors = 256314854,256265573,256257520,256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
> Do you still believe that the speclet "produces cases that require that we generate illegal IL"?

No, I now believe that the speclet describes an implementation that fails to implement the language specification.  The only accessibility requirement on a member accessed through an explicit base is the existing spec constraint that a member found through lookup is accessible.  In this case the member found (the original interface definition) is indeed accessible.  The constraint "The most specific implementation must be accessible at the call site." is one that doesn't appear in the language specification.  It wouldn't even make sense there, as explicit overrides do not have accessibility in the language spec and are not found during name lookup.

In short, unit test `ExplicitBase_152` has errors that are not supported by the language.

---
In reply to: [256494259](https://github.com/dotnet/roslyn/pull/33281#discussion_r256494259) [](ancestors = 256494259,256314854,256265573,256257520,256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)

Reply:
I am glad this is resolved now. And thank you for sharing your thoughts about the feature design.

---
In reply to: [256640387](https://github.com/dotnet/roslyn/pull/33281#discussion_r256640387) [](ancestors = 256640387,256494259,256314854,256265573,256257520,256257340,256177815,255803908,255803140,255772556,255770872,255770151,255762670,255757955)